File: daily/cpp/1578_min_time_to_make_rope_colorful.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
        int n = colors.size();
        vector<int> dp(n, -1);
        return calMinCost(n-1, colors, neededTime, dp, 'A', neededTime[n-1]);
    }

private:
    int calMinCost(int i, string& colors, vector<int>& neededTime, vector<int>& memo, char prevColor, int prevTime) {
        if (i < 0) return 0; // base case: beginning of the string
        if (memo[i] != -1) return memo[i]; // memoization, calculated before
        if (colors[i] == prevColor) {
            return memo[i] = calMinCost(i-1, colors, neededTime, memo, colors[i], max(neededTime[i], prevTime)) + min(neededTime[i], prevTime);
        } else {
            return memo[i] = calMinCost(i-1, colors, neededTime, memo, colors[i], neededTime[i]);
        }
    }
};

File: daily/rust/1578_min_time_to_make_rope_colorful.rs
impl Solution {
    pub fn min_cost(colors: String, needed_time: Vec<i32>) -> i32 {
        let n = colors.len();
        let mut dp = vec![-1; n];
        Self::cal_min_cost(n-1, &colors, needed_time, &mut dp, 'A', needed_time[n-1])
    }

    fn cal_min_cost(i: i32, colors: &String, needed_time: &Vec<i32>, memo: Vec<i32>, prev_color: char, prev_time: i32) -> i32 {
        if i < 0 { return 0; } // base case: return beginning sequence
        if memo[i] != -1 { return memo[i]; } // memoization: return if subproblem already solved
        if colors[i as usize] == prev_color { // if curr color is same as prev color
            // memo[i] = Self::cal_min_cost(i-1, colors, needed_time, memo, colors[i as usize], prev_time.min(needed_time[i as usize]) + 1);
            return memo[i];
        } else { // if curr color diffs 
            memo[i] = Self::cal_min_cost(i-1, colors, needed_time, memo, colors[i as usize], needed_time[i as usize]);
            return memo[i];
        }
    }
}